README.md
<!-- header:start -->
<h1 align="center">
  Basic Temperature
</h1>
<!-- header:end -->

<!-- badges:start -->
[![Gem Version](https://badge.fury.io/rb/basic_temperature.svg)](https://rubygems.org/gems/basic_temperature) [![GitHub Actions CI](https://github.com/marian13/basic_temperature/actions/workflows/ci.yml/badge.svg?branch=main)](https://github.com/marian13/basic_temperature/actions/workflows/ci.yml) [![Maintainability](https://api.codeclimate.com/v1/badges/21dc5d50cf5de8346a3c/maintainability)](https://codeclimate.com/github/marian13/basic_temperature/maintainability) [![Ruby Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://github.com/testdouble/standard) [![Coverage Status](https://coveralls.io/repos/github/marian13/basic_temperature/badge.svg)](https://coveralls.io/github/marian13/basic_temperature?branch=main) [![Inline docs](http://inch-ci.org/github/marian13/basic_temperature.svg?branch=master)](http://inch-ci.org/github/marian13/basic_temperature)
[![Patreon](https://img.shields.io/badge/patreon-donate-orange.svg)](https://www.patreon.com/user?u=31435716&fan_landing=true)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
<!-- badges:end -->

<!-- logo:start -->
<p align="center">
  <img src="https://raw.githubusercontent.com/marian13/basic_temperature/master/logo.png">
</p>
<!-- logo:end -->

<!-- general_description:start -->
`basic_temperature` is a Ruby library which provides a simple [value object](https://martinfowler.com/bliki/ValueObject.html) to work with temperatures and 
allows to perform basic operations like conversion from Celsius to Kelvin, from Kelvin to Fahrenheit etc.
<!-- general_description:end -->

<!-- features:start -->
### Features
- Provides a `Temperature` class which encapsulates all information about a certain  temperature, such
as its amount of degrees and its scale.
- Provides APIs for exchanging temperatures from one scale to another (currently Celsius, Fahrenheit, Kelvin and
Rankine).

- Allows comparing temperatures between each other.
- Queries like `boil_water?`, `freeze_water?`.
- Tested against Ruby 2.3, 2.4, 2.5, 2.6, 2.7, 3.0 and 3.1. See [.github/workflows/ci.yml](https://github.com/marian13/basic_temperature/blob/main/.github/workflows/ci.yml).
<!-- features:end -->

<!-- dependencies:start -->
### Dependencies

* None.
<!-- dependencies:end -->

<!-- documentation:start -->
### Documentation

* Visit https://marian13.github.io/basic_temperature/ to view the documentation.
<!-- documentation:end -->

<!-- installation:start -->
## Installation

Gemfile:

```ruby
gem 'basic_temperature', '~> 1.0.0'
```

And then run:

    $ bundle install

And that's it.

You can access all the features of `basic_temperature` by creating instances of `BasicTemperature::Temperature`.

But there is a shorter form.

If `Temperature` constant was not used before in your app, you can add this line to your Gemfile:

```ruby
gem 'basic_temperature', '~> 1.0.0', require: ['basic_temperature/alias']
```

This way `BasicTemperature::Temperature` class will be accesible simply by `Temperature`.

The following guide assumes you have chosen the shorter form.

If not, just replace all `Temperature` to `BasicTemperature::Temperature`.
<!-- installation:end -->

<!-- usage:start -->
## Usage

### Creating Temperatures

A new temperature can be created in multiple ways:

- Using keyword arguments:

```ruby
Temperature.new(degrees: 0, scale: :celsius)
```

- Using positional arguments:

```ruby
Temperature.new(0, :celsius)
```

- Even more concise way using `Temperature.[]` (an alias of `Temperature.new`):

```ruby
Temperature[0, :celsius]
```

### Creating Temperatures from already existing temperature objects

Sometimes it is useful to create a new temperature from an already existing one.

For such cases, there are `set_degrees` and `set_scale`.

Since temperatures are [value objects](https://martinfowler.com/bliki/ValueObject.html), both methods return
new instances.

Examples:

```ruby
temperature = Temperature[0, :celsius]
# => 0 °C

new_temperature = temperature.set_degrees(15)
# => 15 °C

temperature = Temperature[0, :celsius]
# => 0 °C

new_temperature = temperature.set_scale(:kelvin)
# => 0 K
```

### Conversions

Temperatures can be converted to different scales.

Currently, the following scales are supported: `Celsius`, `Fahrenheit`, `Kelvin` and `Rankine`.

```ruby
Temperature[20, :celsius].to_celsius
# => 20 °C

Temperature[20, :celsius].to_fahrenheit
# => 68 °F

Temperature[20, :celsius].to_kelvin
# => 293.15 K

Temperature[20, :celsius].to_rankine
# => 527.67 °R
```

If it is necessary to convert scale dynamically, `to_scale` method is available.

```ruby
Temperature[20, :celsius].to_scale(scale)
```

All conversion formulas are taken from
[RapidTables](https://www.rapidtables.com/convert/temperature/index.html).

Conversion precision: 2 accurate digits after the decimal dot.

### Comparison

Temperature implements idiomatic [<=> spaceship operator](https://ruby-doc.org/core/Comparable.html) and mixes in [Comparable](https://ruby-doc.org/core/Comparable.html) module.

As a result, all methods from [Comparable](https://ruby-doc.org/core/Comparable.html) are available, e.g:

```ruby
Temperature[20, :celsius] < Temperature[25, :celsius]
# => true

Temperature[20, :celsius] <= Temperature[25, :celsius]
# => true

Temperature[20, :celsius] == Temperature[25, :celsius]
# => false

Temperature[20, :celsius] > Temperature[25, :celsius]
# => false

Temperature[20, :celsius] >= Temperature[25, :celsius]
# => false

Temperature[20, :celsius].between?(Temperature[15, :celsius], Temperature[25, :celsius])
# => true

# Starting from Ruby 2.4.6
Temperature[20, :celsius].clamp(Temperature[20, :celsius], Temperature[25, :celsius])
# => 20 °C
```

Please note, if the second temperature has a different scale, the first temperature is automatically
converted to that scale before comparison.

```ruby
Temperature[20, :celsius] == Temperature[293.15, :kelvin]
# => true
```

#### IMPORTANT !!!

`degrees` are rounded to the nearest value with a precision of 2 decimal digits before comparison.

This means the following temperatures are considered as equal:

```ruby
Temperature[20.020, :celsius] == Temperature[20.024, :celsius]
# => true

Temperature[20.025, :celsius] == Temperature[20.029, :celsius]
# => true
```

while these ones are treated as NOT equal:

```ruby
Temperature[20.024, :celsius] == Temperature[20.029, :celsius]
# => false
```

### Queries
```ruby
Temperature[0, :celsius].boil_water?
# => false

Temperature[0, :celsius].freeze_water?
# => true
```
<!-- usage:end -->

<!-- versioning:start -->
## Versioning
Basic Temperature follows the [Semantic Versioning](https://semver.org/) standard.
<!-- versioning:end -->

<!-- contributing:start -->
## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/marian13/basic_temperature.
<!-- contributing:end -->

<!-- development:start -->
## Development

### Local Machine

* Check specs:
  ```bash
  bundle exec rspec
  ```

* Check linter:
  ```bash
  bundle exec rubocop
  ```

* Update docs:
  ```bash
  bundle exec sdoc lib -T rails -o docs
  ```

* Find missed docs:
  ```bash
  bundle exec inch
  ```

* Check for vulnerable gem versions, insecure gem sources, etc...
  ```bash
  bundle exec bundle-audit check --update
  ```

### Docker

* Build container:
  ```bash
  docker build . -f docker/3.1/Dockerfile -t basic_temperature:3.1
  ```

* Run specs inside the container:
  ```bash
  docker run --rm -it basic_temperature:3.1 bundle exec rspec
  ```

* Interactive Bash shell:
  ```bash
  docker run --rm -it basic_temperature:3.1 bash
  ```

* Ruby REPL with already required gem files:
  ```bash
  docker run --rm -it basic_temperature:3.1 bin/console
  ```

* `3.1` can be replaced by `2.3`, `2.4`, `2.5`, `2.6`, `2.7`, and `3.0`.
<!-- development:end -->

<!-- license:start -->
## License

The gem is available as open-source under the terms of the [MIT License](https://opensource.org/licenses/MIT).
<!-- license:end -->

<!-- author:start -->
Copyright (c) 2022 [Marian Kostyk](http://mariankostyk.com).
<!-- author:end -->


CHANGELOG.md
## 1.0.0

### Features:
  - Add `Temperature#boil_water?` and `Temperature#freeze_water?` (#47).

### Removals:
  - Remove `BasicTemperature#==`, since it is generated by `Comparable` (#43).
  - Remove `Math`, since it provides more confusion than benefit (#51).

### Docs:
  - Add link to docs in `README.md` (#42).

### Tech improvements:
  - Split [God Object](https://sourcemaking.com/antipatterns/the-blob) into modules (#43) .
  - Set up [Inch CI](https://inch-ci.org/) (#45).

## 0.2.2

### Features:
  - Include [Comparable](https://ruby-doc.org/core-3.0.1/Comparable.html) (#24).
  - Introduce short form: `Temperature` instead of `BasicTemperature::Temperature` (#25).
  - Add simple memoization (#26).
  - Add rounding (#26).
  - Introduce alias `[]` for `new` (#27).
  - Add support of `Rankine` scale.
  - Implement `Temperature#inspect`.

### Fixes:
  - Update `BasicTemperature#==` (When two temperatures had different scales - automatic conversion was not performed) (#18).
  - Restrict `Rubocop` version for `Code Climate` (#22).
  - Fix typo `Celsium` instead of `Celcius` (#23).
  - Add `Gemfile.lock` to `.gitignore` (#38).

### Docs
  - Add `Active Development Warning` (#19).
  - Introduce badges: `Gem Version`, `Build Status`, `Maintainability`, `Coverage Status`, `License: MIT`, `Patreon`, `Inch CI` (#14) (#20) (#40).
  - Add docs for all public API (#38).
  - Add usage section in `README.md` (#38).

### Tech improvements:
  - Set up [Coverals](https://docs.coveralls.io/ruby-on-rails) (#10).
  - Set up [SimpleCov](https://github.com/colszowka/simplecov) (#10).
  - Set up [Code Climate](https://codeclimate.com/) (#13).
  - Set up [ReverseCoverage] (#21).
  - Set up [SDoc] (#38).
  - Set up [Rerun] (#38).
  - Autodeploy docs to [Github Pages](https://pages.github.com/) (#39).

## 0.2.1

### Features:
  - Add spaceship operator [<=>](https://dev.to/jblengino510/the-spaceship-operator-in-ruby-2fmk) (#2).

### Tech improvements:
  - Set up [Byebug](https://github.com/deivid-rodriguez/byebug) (#1).
  - Set up [RSpec](https://rspec.info/) (#3).
  - Set up [RuboCop](https://github.com/rubocop-hq/rubocop) (#5).
  - Set up [Travis CI](https://travis-ci.com/) (#4).

## 0.2.0
  - 'Accidentally' yanked.

## 0.1.0
  - Initial Release.

### Features:
  - Create `Temperature` using positional arguments.
  - Create `Temperature` using keyword arguments.
  - Create `Temperature`s from already existing temperature objects (`Temperature#set_degrees`, `Temperature#set_scale`).
  - Conversion to `Celsius`, `Fahrenheit`, `Kelvin`.
  - Add dynamic conversion (`Temperature#to_scale`).
  - Introduce equality operator `==`.
  - Implement addition of `Temperature`s.
  - Implement subtraction of `Temperature`s.
  - Support Ruby coersion mechanism.


lib/basic_temperature/temperature.rb
# frozen_string_literal: true

require_relative "temperature/additional_helpers"
require_relative "temperature/assertions"
require_relative "temperature/casting"
require_relative "temperature/errors"
require_relative "temperature/initialization"
require_relative "temperature/memoization"
require_relative "temperature/rounding"

module BasicTemperature
  ##
  # Temperature is a simple {Value Object}[https://martinfowler.com/bliki/ValueObject.html] for basic
  # temperature operations like conversions from <tt>Celsius</tt> to <tt>Fahrenhait</tt> or <tt>Kelvin</tt>
  # etc.
  #
  # Supported scales: <tt>Celsius</tt>, <tt>Fahrenheit</tt>, <tt>Kelvin</tt> and <tt>Rankine</tt>.
  #
  # == Creating Temperatures
  #
  # A new temperature can be created in multiple ways:
  #
  # - Using keyword arguments:
  #
  #     Temperature.new(degrees: 0, scale: :celsius)
  #
  # - Using positional arguments:
  #
  #     Temperature.new(0, :celsius)
  #
  # - Even more concise way using <tt>Temperature.[]</tt> (an alias of <tt>Temperature.new</tt>):
  #
  #     Temperature[0, :celsius]
  #
  #
  # == Creating Temperatures from already existing temperature objects
  #
  # Sometimes it is useful to create a new temperature from already existing one.
  #
  # For such cases, there are {set_degrees}[rdoc-ref:Temperature#set_degrees] and
  # {set_scale}[rdoc-ref:Temperature#set_scale].
  #
  # Since temperatures are {Value Objects}[https://martinfowler.com/bliki/ValueObject.html], both methods
  # returns new instances.
  #
  # Examples:
  #
  #   temperature = Temperature[0, :celsius]
  #   # => 0 °C
  #
  #   new_temperature = temperature.set_degrees(15)
  #   # => 15 °C
  #
  #   temperature = Temperature[0, :celsius]
  #   # => 0 °C
  #
  #   new_temperature = temperature.set_scale(:kelvin)
  #   # => 0 K
  #
  # == Conversions
  #
  # Temperatures can be converted to diffirent scales.
  #
  # Currently, the following scales are supported: <tt>Celsius</tt>, <tt>Fahrenheit</tt>, <tt>Kelvin</tt> and
  # <tt>Rankine</tt>.
  #
  #   Temperature[20, :celsius].to_celsius
  #   # => 20 °C
  #
  #   Temperature[20, :celsius].to_fahrenheit
  #   # => 68 °F
  #
  #   Temperature[20, :celsius].to_kelvin
  #   # => 293.15 K
  #
  #   Temperature[20, :celsius].to_rankine
  #   # => 527.67 °R
  #
  # If it is necessary to convert scale dynamically, {to_scale}[rdoc-ref:Temperature#to_scale] method is
  # available.
  #
  #   Temperature[20, :celsius].to_scale(scale)
  #
  # All conversion formulas are taken from
  # {RapidTables}[https://www.rapidtables.com/convert/temperature/index.html].
  #
  # Conversion precision: 2 accurate digits after the decimal dot.
  #
  # == Comparison
  #
  # Temperature implements idiomatic {<=> spaceship operator}[https://ruby-doc.org/core/Comparable.html] and
  # mixes in {Comparable}[https://ruby-doc.org/core/Comparable.html] module.
  #
  # As a result, all methods from Comparable are available, e.g:
  #
  #   Temperature[20, :celsius] < Temperature[25, :celsius]
  #   # => true
  #
  #   Temperature[20, :celsius] <= Temperature[25, :celsius]
  #   # => true
  #
  #   Temperature[20, :celsius] == Temperature[25, :celsius]
  #   # => false
  #
  #   Temperature[20, :celsius] > Temperature[25, :celsius]
  #   # => false
  #
  #   Temperature[20, :celsius] >= Temperature[25, :celsius]
  #   # => false
  #
  #   Temperature[20, :celsius].between?(Temperature[15, :celsius], Temperature[25, :celsius])
  #   # => true
  #
  #   # Starting from Ruby 2.4.6
  #   Temperature[20, :celsius].clamp(Temperature[20, :celsius], Temperature[25, :celsius])
  #   # => 20 °C
  #
  # Please note, if <tt>other</tt> temperature has a different scale, temperature is automatically converted
  # to that scale before comparison.
  #
  #   Temperature[20, :celsius] == Temperature[293.15, :kelvin]
  #   # => true
  #
  # IMPORTANT !!!
  #
  # <tt>degrees</tt> are rounded to the nearest value with a precision of 2 decimal digits before comparison.
  #
  # This means the following temperatures are considered as equal:
  #
  #   Temperature[20.020, :celsius] == Temperature[20.024, :celsius]
  #   # => true
  #
  #   Temperature[20.025, :celsius] == Temperature[20.029, :celsius]
  #   # => true
  #
  # while these ones are treated as NOT equal:
  #
  #   Temperature[20.024, :celsius] == Temperature[20.029, :celsius]
  #   # => false
  #
  # == Math
  #
  # ==== Addition/Subtraction.
  #
  #   Temperature[20, :celsius] + Temperature[10, :celsius]
  #   # => 30 °C
  #
  #   Temperature[20, :celsius] - Temperature[10, :celsius]
  #   # => 10 °C
  #
  # If second temperature has a different scale, first temperature is automatically converted to that scale
  # before <tt>degrees</tt> addition/subtraction.
  #
  #   Temperature[283.15, :kelvin] + Temperature[10, :celsius]
  #   # => 10 °C
  #
  # Returned temperature will have the same scale as the second temperature.
  #
  # It is possible to add/subtract numerics.
  #
  #   Temperature[20, :celsius] + 10
  #   # => 30 °C
  #
  #   Temperature[20, :celsius] - 10
  #   # => 10 °C
  #
  # In such cases, returned temperature will have the same scale as the first temperature.
  #
  # Also {Ruby coersion mechanism}[https://ruby-doc.org/core/Numeric.html#method-i-coerce] is supported.
  #
  #   10 + Temperature[20, :celsius]
  #   # => 30 °C
  #
  #   10 - Temperature[20, :celsius]
  #   # => -10 °C
  #
  # ==== Negation
  #
  #   -Temperature[20, :celsius]
  #   # => -20 °C
  #
  # == Queries
  #
  #   Temperature[0, :celsius].boil_water?
  #   # => false
  #
  #   Temperature[0, :celsius].freeze_water?
  #   # => true
  #
  class Temperature
    include Comparable

    include AdditionalHelpers
    include Assertions
    include Casting
    include Errors
    include Initialization
    include Memoization
    include Rounding

    CELSIUS = "celsius"
    FAHRENHEIT = "fahrenheit"
    KELVIN = "kelvin"
    RANKINE = "rankine"

    # A list of all currently supported scale values.
    SCALES = [CELSIUS, FAHRENHEIT, KELVIN, RANKINE].freeze

    # Degrees of the temperature.
    attr_reader :degrees

    # Scale of the temperature. Look at {SCALES}[rdoc-ref:Temperature::SCALES] for possible values.
    attr_reader :scale

    ##
    # Creates a new instance of Temperature. Alias for <tt>new</tt>.
    #
    # :call-seq:
    #   [](degrees:, scale:)
    #   [](degrees, scale)
    #
    def self.[](*args, **kwargs)
      new(*args, **kwargs)
    end

    ##
    # Creates a new instance of Temperature. Is aliased as <tt>[]</tt>.
    #
    # :call-seq:
    #   new(degrees:, scale:)
    #   new(degrees, scale)
    #
    def initialize(*positional_arguments, **keyword_arguments)
      assert_either_positional_arguments_or_keyword_arguments!(positional_arguments, keyword_arguments)

      if keyword_arguments.any?
        initialize_via_keywords_arguments(keyword_arguments)
      else # positional_arguments.any?
        initialize_via_positional_arguments(positional_arguments)
      end
    end

    # Returns a new Temperature with updated <tt>degrees</tt>.
    #
    #   temperature = Temperature[0, :celsius]
    #   # => 0 °C
    #
    #   new_temperature = temperature.set_degrees(15)
    #   # => 15 °C
    #
    def set_degrees(degrees)
      Temperature.new(degrees, scale)
    end

    # Returns a new Temperature with updated <tt>scale</tt>.
    #
    #   temperature = Temperature[0, :celsius]
    #   # => 0 °C
    #
    #   new_temperature = temperature.set_scale(:kelvin)
    #   # => 0 K
    #
    def set_scale(scale)
      Temperature.new(degrees, scale)
    end

    ##
    # Converts temperature to specific <tt>scale</tt>.
    # If temperature is already in desired <tt>scale</tt>, returns current temperature object.
    #
    # Raises {InvalidScaleError}[rdoc-ref:Temperature::InvalidScaleError]
    # when <tt>scale</tt> can not be casted to any possible scale value
    # (see {SCALES}[rdoc-ref:Temperature::SCALES]).
    #
    #   Temperature[60, :fahrenheit].to_scale(:celsius)
    #   # => 15.56 °C
    #
    def to_scale(scale)
      casted_scale = cast_scale(scale)

      assert_valid_scale!(casted_scale)

      case casted_scale
      when CELSIUS
        to_celsius
      when FAHRENHEIT
        to_fahrenheit
      when KELVIN
        to_kelvin
      when RANKINE
        to_rankine
      end
    end

    ##
    # Converts temperature to Celsius scale. If temperature is already in Celsius, returns current
    # temperature object.
    #
    # Memoizes subsequent calls.
    #
    # Conversion formulas are taken from {RapidTables}[https://www.rapidtables.com/]:
    # 1. {Celsius to Fahrenheit}[https://www.rapidtables.com/convert/temperature/celsius-to-fahrenheit.html].
    # 2. {Celsius to Kelvin}[https://www.rapidtables.com/convert/temperature/celsius-to-kelvin.html].
    # 3. {Celsius to Rankine}[https://www.rapidtables.com/convert/temperature/celsius-to-rankine.html].
    #
    #   Temperature[0, :fahrenheit].to_celsius
    #   # => -17.78 °C
    #
    def to_celsius
      memoized(:to_celsius) || memoize(:to_celsius, -> {
        return self if scale == CELSIUS

        degrees =
          case scale
          when FAHRENHEIT
            (self.degrees - 32) * (5 / 9r)
          when KELVIN
            self.degrees - 273.15
          when RANKINE
            (self.degrees - 491.67) * (5 / 9r)
          end

        Temperature.new(degrees, CELSIUS)
      })
    end

    ##
    # Converts temperature to Fahrenheit scale. If temperature is already in Fahrenheit, returns current
    # temperature object.
    #
    # Memoizes subsequent calls.
    #
    # Conversion formulas are taken from {RapidTables}[https://www.rapidtables.com/]:
    # 1. {Fahrenheit to Celsius}[https://www.rapidtables.com/convert/temperature/fahrenheit-to-celsius.html].
    # 2. {Fahrenheit to Kelvin}[https://www.rapidtables.com/convert/temperature/fahrenheit-to-kelvin.html].
    # 3. {Fahrenheit to Rankine}[https://www.rapidtables.com/convert/temperature/fahrenheit-to-rankine.html].
    #
    #   Temperature[0, :celsius].to_fahrenheit
    #   # => 32 °F
    #
    def to_fahrenheit
      memoized(:to_fahrenheit) || memoize(:to_fahrenheit, -> {
        return self if scale == FAHRENHEIT

        degrees =
          case scale
          when CELSIUS
            self.degrees * (9 / 5r) + 32
          when KELVIN
            self.degrees * (9 / 5r) - 459.67
          when RANKINE
            self.degrees - 459.67
          end

        Temperature.new(degrees, FAHRENHEIT)
      })
    end

    ##
    # Converts temperature to Kelvin scale. If temperature is already in Kelvin, returns current
    #  temperature object.
    #
    # Memoizes subsequent calls.
    #
    # Conversion formulas are taken from {RapidTables}[https://www.rapidtables.com/]:
    # 1. {Kelvin to Celsius}[https://www.rapidtables.com/convert/temperature/kelvin-to-celsius.html].
    # 2. {Kelvin to Fahrenheit}[https://www.rapidtables.com/convert/temperature/kelvin-to-fahrenheit.html].
    # 3. {Kelvin to Rankine}[https://www.rapidtables.com/convert/temperature/kelvin-to-rankine.html].
    #
    #   Temperature[0, :kelvin].to_rankine
    #   # => 0 °R
    #
    def to_kelvin
      memoized(:to_kelvin) || memoize(:to_kelvin, -> {
        return self if scale == KELVIN

        degrees =
          case scale
          when CELSIUS
            self.degrees + 273.15
          when FAHRENHEIT
            (self.degrees + 459.67) * (5 / 9r)
          when RANKINE
            self.degrees * (5 / 9r)
          end

        Temperature.new(degrees, KELVIN)
      })
    end

    ##
    # Converts temperature to Rankine scale. If temperature is already in Rankine, returns current
    # temperature object.
    #
    # Memoizes subsequent calls.
    #
    # Conversion formulas are taken from {RapidTables}[https://www.rapidtables.com/]:
    # 1. {Rankine to Celsius}[https://www.rapidtables.com/convert/temperature/rankine-to-celsius.html].
    # 2. {Rankine to Fahrenheit}[https://www.rapidtables.com/convert/temperature/rankine-to-fahrenheit.html].
    # 3. {Rankine to Kelvin}[https://www.rapidtables.com/convert/temperature/rankine-to-kelvin.html].
    #
    #   Temperature[0, :rankine].to_kelvin
    #   # => 0 K
    #
    def to_rankine
      memoized(:to_rankine) || memoize(:to_rankine, -> {
        return self if scale == RANKINE

        degrees =
          case scale
          when CELSIUS
            (self.degrees + 273.15) * (9 / 5r)
          when FAHRENHEIT
            self.degrees + 459.67
          when KELVIN
            self.degrees * (9 / 5r)
          end

        Temperature.new(degrees, RANKINE)
      })
    end

    ##
    # Compares temperture with <tt>other</tt> temperature.
    #
    # Returns <tt>0</tt> if they are considered as equal.
    #
    # Two temperatures are considered as equal when they have the same amount of <tt>degrees</tt>.
    #
    # Returns <tt>-1</tt> if temperature is lower than <tt>other</tt> temperature.
    #
    # Returns <tt>1</tt> if temperature is higher than <tt>other</tt> temperature.
    #
    # If <tt>other</tt> temperature has a different scale, temperature is automatically converted to that
    # scale before <tt>degrees</tt> comparison.
    #
    #   Temperature[20, :celsius] <=> Temperature[20, :celsius]
    #   # => 0
    #
    #   Temperature[20, :celsius] <=> Temperature[293.15, :kelvin]
    #   # => 0
    #
    # IMPORTANT!!!
    #
    # This method rounds <tt>degrees</tt> to the nearest value with a precision of 2 decimal digits.
    #
    # This means the following:
    #
    #   Temperature[20.020, :celsius] <=> Temperature[20.024, :celsius]
    #   # => 0
    #
    #   Temperature[20.025, :celsius] <=> Temperature[20.029, :celsius]
    #   # => 0
    #
    #   Temperature[20.024, :celsius] <=> Temperature[20.029, :celsius]
    #   # => -1
    #
    def <=>(other)
      return unless assert_temperature(other)

      round_degrees(to_scale(other.scale).degrees) <=> round_degrees(other.degrees)
    end

    ##
    # Returns true when temperature boils water (is greater than or equal to 100 °C),
    # false otherwise.
    #
    def boil_water?
      to_celsius.degrees >= 100
    end

    ##
    # Returns true when temperature freezes water (is less than or equal to 0 °C),
    # false otherwise.
    #
    def freeze_water?
      to_celsius.degrees <= 0
    end

    # Is used by {+}[rdoc-ref:Temperature#+] and {-}[rdoc-ref:Temperature#-]
    # for {Ruby coersion mechanism}[https://ruby-doc.org/core/Numeric.html#method-i-coerce].
    def coerce(numeric) # :nodoc:
      assert_numeric!(numeric)

      [Temperature.new(numeric, scale), self]
    end

    # Returns a string containing a human-readable representation of temperature.
    def inspect # :nodoc:
      rounded_degrees = round_degrees(degrees)

      printable_degrees = degrees_without_decimal?(rounded_degrees) ? rounded_degrees.to_i : rounded_degrees

      scale_symbol =
        case scale
        when CELSIUS
          "°C"
        when FAHRENHEIT
          "°F"
        when KELVIN
          "K"
        when RANKINE
          "°R"
        end

      "#{printable_degrees} #{scale_symbol}"
    end
  end
  # rubocop:enable Metrics/ClassLength
end
